src/SignalR/clients/ts/signalr/src/MessageBuffer.ts

```diff
@@ -0,0 +1,218 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { IConnection } from "./IConnection";
+import { AckMessage, HubMessage, IHubProtocol, MessageType, SequenceMessage } from "./IHubProtocol";
+import { isArrayBuffer } from "./Utils";
+
+/** @private */
+export class MessageBuffer {
+    private readonly _protocol: IHubProtocol;
+    private readonly _connection: IConnection;
+
+    private readonly _bufferSize: number = 100_000;
+
+    private _messages: Key[] = [];
+    private _totalMessageCount: number = 0;
+    private _waitForSequenceMessage: boolean = false;
+
+    // Message IDs start at 1 and always increment by 1
+    private _currentReceivingSequenceId = 1;
+    private _latestReceivedSequenceId = -100;
+    private _bufferedByteCount: number = 0;
+
+    private _ackTimerHandle?: any;
+
+    private _backPressurePromise!: Promise<void>;
+    private _backpressurePromiseResolver: (value: void) => void = () => {};
+
+    private _resendPromise: Promise<void> = Promise.resolve();
+    private _resendResolve?: (value: void) => void = undefined;
+    private _resendReject?: (value?: any) => void = undefined;
+
+    constructor(protocol: IHubProtocol, connection: IConnection, bufferSize: number) {
+        this._protocol = protocol;
+        this._connection = connection;
+        this._bufferSize = bufferSize;
+    }
+
+    public async _send(message: HubMessage): Promise<void> {
+        if (this._bufferedByteCount >= this._bufferSize) {
+            await this._backPressurePromise;
+        }
+
+        // If the connection is down we want to wait for it to come back or fail to reconnect
+        await this._resendPromise;
```

Comment:
Does it make sense for everything that gets sent during the stateful reconnect period to bypass the `_backPressurePromise` for the new connection after everything is resent? Of course, awaiting the `_resendPromise` first seems problematic too because of ordering issues. I know that promises are guaranteed to call awaiters in the order they were added, but multiple promises do make it harder to reason about.

I wonder if we shouldn't just immediately put the `HubMessage` in `_messages` before awaiting anything so we have a single source of truth for message order that never changes based on when various promise complete. It would be guaranteed to be the same whether or not you enable stateful reconnect.

We could have a single loop that reads out of the unsent `_messages` and calls `_connection.send` in a single loop. We could put the promise resolver in the `Key` when there should be backpressure. And then `_ack` can search past the most recent acked message and resolve promises if they're close enough to fall under the `_bufferSize`.

Trying to think about what happens when there are multiple callers awaiting `_backPressurePromise` and/or `_resendPromise` inside this method at the same time is making my brain hurt.